ocaml/xapi/xapi_database_backup.ml

```diff
@@ -0,0 +1,184 @@
+module D=Debug.Make(struct let name="xapi_database_backup" end)
+open D
+
+open Http
+open Xmlrpc_client
+open Db_cache_types
+open Stdext
+open Threadext
+
+type stat  = {created : int64; modified : int64; deleted : int64;    } [@@deriving rpc]
+type field = {fldname : string; stat : stat; value : Schema.Value.t; } [@@deriving rpc]
+type row   = {objref  : string; stat : stat; value : field list;     } [@@deriving rpc]
+type table = {tblname : string; stat : stat; value : row list;       } [@@deriving rpc]
+
+type del_tables = {key: string; table: string; stat : stat; } [@@deriving rpc]
+type delta = {fresh_token: int64; tables : table list; deletes : del_tables list; } [@@deriving rpc]
+type edits = {ts : TableSet.t; tables : string list}
+
+type count = {tblname:string; count: int; del:int;}
+
+let empty_schema = Schema.Value.unmarshal String ""
+let empty_field = {fldname=""; stat={created=0L;modified=0L;deleted=0L;}; value=empty_schema;}
+let empty_row   = {objref =""; stat={created=0L;modified=0L;deleted=0L;}; value=[empty_field];}
+
+
+exception Content_length_required
+
+let delta_to_string (d:delta) = Jsonrpc.to_string (rpc_of_delta d)
+
+(* Master Functions *)
+
+let get_del_from_table_name token ts table =
+  Table.fold_over_deleted token (fun key {created; modified; deleted} acc ->
+      ({key; table; stat={created;modified;deleted}})::acc) (TableSet.find table ts) []
+
+let get_deleted ~__context ~token db =
+  let tables =
+    let all =
+      let objs = List.filter (fun x -> x.Datamodel_types.gen_events) (Dm_api.objects_of_api Datamodel.all_api) in
+      List.map (fun x->x.Datamodel_types.name) objs in
```

Comment:
`List.map` is not tail recursive, it can overflow the stack if the list is large. There's an alternative one in `xapi-stdext` .
If we move to 4.07 this would be a good candidate for use of `Seq` though (or we can start using `seq` with 4.06 if we add the library I think).

Reply:
I'm not sure that seq is correct for this? Wouldn't the benefit of Seq be from a list that isn't fully created? I already have the full list in the reply from `(Dm_api.objects_of_api Datamodel.all_api)` , or am I missing something